docs/compilation/performance_tips.md

```diff
@@ -0,0 +1,645 @@
+4 bits ->    61_979_934 complexity -> 0.021961s
+5 bits ->    99_198_195 complexity -> 0.029475s
+6 bits ->   230_210_706 complexity -> 0.062841s
+7 bits ->   535_706_740 complexity -> 0.139669s
+8 bits -> 1_217_510_420 complexity -> 0.318838s
+```
+
+And displays:
+![](../_static/compilation/performance_tips/complexity_and_timing_per_bit_width.png)
+
+Here are some ways to optimize table lookups.
+
+### By reducing the amount of table lookups
+
+This one is probably the most complicated one in this list as it's not automated. The idea is to use mathematical properties of operations to reduce the amount of table lookups needed to achieve the result.
+
+One great example is in adding big integers in bitmap representation. Here is the basic implementation:
+
+```python
+def add_bitmaps(x, y):
+    result = fhe.zeros((N,))
+    carry = 0
+
+    addition = x + y
+    for i in range(N):
+        addition_and_carry = addition[i] + carry
+        carry = addition_and_carry >> 1
+        result[i] = addition_and_carry % 2
+
+    return result
+```
+
+There are two table lookups within the loop body, one for `>>` and one for `%`.
+
+This implementation is not optimal though, since the same output can be achieved with just a single table lookup:
+
+```python
+def add_bitmaps(x, y):
+    result = fhe.zeros((N,))
+    carry = 0
+
+    addition = x + y
+    for i in range(N):
+        addition_and_carry = addition[i] + carry
+        carry = addition_and_carry >> 1
+        result[i] = addition_and_carry - (carry * 2)
+
+    return result
+```
+
+It was possible to do this because the original operations had a mathematical equivalence with the optimized operations and optimized operations achieved the same output with less table lookups!
+
+Here is the full code example and some numbers for this optimization:
+
+```python
+import numpy as np
+from concrete import fhe
+
+N = 32
+
+def add_bitmaps_naive(x, y):
+    result = fhe.zeros((N,))
+    carry = 0
+
+    addition = x + y
+    for i in range(N):
+        addition_and_carry = addition[i] + carry
+        carry = addition_and_carry >= 2
+        result[i] = addition_and_carry % 2
+
+    return result
+
+def add_bitmaps_optimized(x, y):
+    result = fhe.zeros((N,))
+    carry = 0
+
+    addition = x + y
+    for i in range(N):
+        addition_and_carry = addition[i] + carry
+        carry = addition_and_carry >> 1
+        result[i] = addition_and_carry - (carry * 2)
+
+    return result
+
+inputset = fhe.inputset(fhe.tensor[fhe.uint1, N], fhe.tensor[fhe.uint1, N])
+for (name, implementation) in [("naive", add_bitmaps_naive), ("optimized", add_bitmaps_optimized)]:
+    compiler = fhe.Compiler(implementation, {"x": "encrypted", "y": "encrypted"})
+    circuit = compiler.compile(inputset)
+
+    print(
+        f"{name:>9} implementation "
+        f"-> {int(circuit.programmable_bootstrap_count)} table lookups "
+        f"-> {int(circuit.complexity):_} complexity"
+    )
+```
+
+prints:
+
+```
```

Comment:
I would reduce, by not giving the full example: it's pretty obvious from the previous explanations that there is half less PBS

Reply:
If the user wants to play with it, it's better to have the full example to just copy-paste. I agree that it's a bit long as is, but if we split, it'd be okay IMO.